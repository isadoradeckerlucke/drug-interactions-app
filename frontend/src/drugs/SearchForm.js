import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { getInteractionsFromAPI } from "../actions/drugs";
import { resetInteractions } from "../actions/resetInteractions";
import { setCurrentUser } from "../actions/currentUser";
import DrugInteractionApi from "../Api";
import { toggleLoading } from "../actions/loading";
import { setSearchTerm } from "../actions/searchTerm";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

function SearchForm() {
  /** form for searching drug interactions - one search bar initially, can add multiple additional drugs */
  const searchTerm = useSelector((st) => st.searchTerm.searchTerm);
  const currentUser = useSelector((st) => st.currentUser);
  const dispatch = useDispatch();
  const [checked, setChecked] = useState(false);

  useEffect(() => {
    dispatch(resetInteractions());
  }, [dispatch]);

  async function handleSubmit(e) {
    e.preventDefault();
    let errorHolder = document.getElementById("error-holder");
    errorHolder.innerText = "";

    if (!searchTerm) {
      errorHolder.innerText = "please enter a search term";
      dispatch(resetInteractions());
      return;
    }

    dispatch(getInteractionsFromAPI(searchTerm.trim()));
    // if it's checked, save the drugs to the database and the user
    // then save the user with the new saved drugs
    let res;
    if (checked) {
      let username = currentUser.currentUser.username;
      if (searchTerm.includes(",")) {
        // check to make sure the drug isn't already saved, if it isn't then add it.
        const nameArray = searchTerm.split(", ");
        for (let name of nameArray) {
          // if they haven't already saved the drug
          if (currentUser.currentUser.savedDrugs.includes(name) !== true) {
            res = await DrugInteractionApi.saveDrugToUser(username, name);
          }
        }
      } else {
        if (currentUser.currentUser.savedDrugs.includes(searchTerm) !== true) {
          res = await DrugInteractionApi.saveDrugToUser(username, searchTerm);
        }
      }
      dispatch(setCurrentUser(username));
    }
    if (!res) {
      dispatch(toggleLoading());
    }
  }

  function handleChange(e) {
    dispatch(setSearchTerm(e.target.value));
  }
  function changeBool(e) {
    setChecked(!checked);
  }

  return (
    <div>
      <h3>
        search for high severity interactions with a single drug, or get all
        interactions between multiple drugs.
      </h3>
      <Form onSubmit={handleSubmit}>
        <Form.Label htmlFor="drugName">search interactions:</Form.Label>
        <Form.Control
          name="drugName"
          placeholder="drug name(s)"
          value={searchTerm}
          onChange={handleChange}
        />
        <br></br>
        {currentUser.isLoggedIn ? (
          <Form.Group controlId="formBasicCheckbox">
            <Form.Check
              type="checkbox"
              label="add searched drugs to saved list"
              name="saveBool"
              value={checked}
              onChange={changeBool}
            />
          </Form.Group>
        ) : (
          <span></span>
        )}
        <Form.Text id="searchHelpBlock" muted>
          please separate multiple drug names by commas
        </Form.Text>
        <br></br>
        <Button type="submit" className="btn btn-lg btn-dark">
          search
        </Button>
        <p id="error-holder"></p>
      </Form>
      <br></br>
    </div>
  );
}

export default SearchForm;
