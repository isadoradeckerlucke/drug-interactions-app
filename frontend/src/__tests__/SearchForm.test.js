import React from "react";
import { render, fireEvent } from "@testing-library/react";
import SearchForm from "../drugs/SearchForm";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

describe("search form works as expected", function () {
  const currentUserTest = {
    isLoggedIn: true,
    currentUser: {
      username: "testuser",
      email: "testuser@gmail.com",
      savedDrugs: [],
    },
  };
  const initialState = {
    searchTerm: { searchTerm: "" },
    currentUser: currentUserTest,
    drugs: { interactions: {} },
    loading: { loading: false },
  };
  const searchTermIn = {
    searchTerm: { searchTerm: "adderall" },
    currentUser: currentUserTest,
    drugs: { interactions: {} },
    loading: { loading: false },
  };
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  let store;

  /** smoke test */
  it("renders without crashing", function () {
    store = mockStore(initialState);
    render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );
  });

  /** snapshot test */
  it("matches initial snapshot", function () {
    store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot with search term in", function () {
    store = mockStore(searchTermIn);

    const { asFragment } = render(
      <Provider store={store}>
        <SearchForm />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
