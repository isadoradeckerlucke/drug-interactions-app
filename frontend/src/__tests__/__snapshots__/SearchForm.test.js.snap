// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search form works as expected matches initial snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      search for high severity interactions with a single drug, or get all interactions between multiple drugs.
    </h3>
    <form
      class=""
    >
      <label
        class="form-label"
        for="drugName"
      >
        search interactions:
      </label>
      <input
        class="form-control"
        name="drugName"
        placeholder="drug name(s)"
        value=""
      />
      <br />
      <div
        class="form-group"
      >
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="formBasicCheckbox"
            name="saveBool"
            type="checkbox"
            value="false"
          />
          <label
            class="form-check-label"
            for="formBasicCheckbox"
            title=""
          >
            add searched drugs to saved list
          </label>
        </div>
      </div>
      <small
        class="form-text text-muted"
        id="searchHelpBlock"
      >
        please separate multiple drug names by commas
      </small>
      <br />
      <button
        class="btn btn-lg btn-dark btn btn-primary"
        type="submit"
      >
        search
      </button>
      <p
        id="error-holder"
      />
    </form>
    <br />
  </div>
</DocumentFragment>
`;

exports[`search form works as expected matches snapshot with search term in 1`] = `
<DocumentFragment>
  <div>
    <h3>
      search for high severity interactions with a single drug, or get all interactions between multiple drugs.
    </h3>
    <form
      class=""
    >
      <label
        class="form-label"
        for="drugName"
      >
        search interactions:
      </label>
      <input
        class="form-control"
        name="drugName"
        placeholder="drug name(s)"
        value="adderall"
      />
      <br />
      <div
        class="form-group"
      >
        <div
          class="form-check"
        >
          <input
            class="form-check-input"
            id="formBasicCheckbox"
            name="saveBool"
            type="checkbox"
            value="false"
          />
          <label
            class="form-check-label"
            for="formBasicCheckbox"
            title=""
          >
            add searched drugs to saved list
          </label>
        </div>
      </div>
      <small
        class="form-text text-muted"
        id="searchHelpBlock"
      >
        please separate multiple drug names by commas
      </small>
      <br />
      <button
        class="btn btn-lg btn-dark btn btn-primary"
        type="submit"
      >
        search
      </button>
      <p
        id="error-holder"
      />
    </form>
    <br />
  </div>
</DocumentFragment>
`;
